ur_controllers/src/ur_configuration_controller.cpp

```diff
@@ -0,0 +1,122 @@
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn
+URConfigurationController::on_configure(const rclcpp_lifecycle::State& /* previous_state */)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration URConfigurationController::command_interface_configuration() const
+{
+  // No command interfaces currently
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  const std::string tf_prefix = params_.tf_prefix;
+
+  return config;
+}
+
+controller_interface::InterfaceConfiguration URConfigurationController::state_interface_configuration() const
+{
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  const std::string tf_prefix = params_.tf_prefix;
+
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_major");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_minor");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_build");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_bugfix");
+
+  return config;
+}
+
+controller_interface::return_type URConfigurationController::update(const rclcpp::Time& /* time */,
+                                                                    const rclcpp::Duration& /* period */)
+{
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::CallbackReturn
+URConfigurationController::on_activate(const rclcpp_lifecycle::State& /* previous_state */)
+{
+  get_robot_software_version_srv_ = get_node()->create_service<ur_msgs::srv::GetRobotSoftwareVersion>(
+      "~/get_robot_software_version", std::bind(&URConfigurationController::getRobotSoftwareVersion, this,
+                                                std::placeholders::_1, std::placeholders::_2));
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn
+URConfigurationController::on_deactivate(const rclcpp_lifecycle::State& /* previous_state */)
+{
+  get_robot_software_version_srv_.reset();
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+bool URConfigurationController::getRobotSoftwareVersion(
+    ur_msgs::srv::GetRobotSoftwareVersion::Request::SharedPtr /*req*/,
+    ur_msgs::srv::GetRobotSoftwareVersion::Response::SharedPtr resp)
+{
+  resp->major = static_cast<uint32_t>(state_interfaces_[StateInterfaces::ROBOT_VERSION_MAJOR].get_value());
+  resp->minor = static_cast<uint32_t>(state_interfaces_[StateInterfaces::ROBOT_VERSION_MINOR].get_value());
+  resp->bugfix = static_cast<uint32_t>(state_interfaces_[StateInterfaces::ROBOT_VERSION_BUGFIX].get_value());
+  resp->build = static_cast<uint32_t>(state_interfaces_[StateInterfaces::ROBOT_VERSION_BUILD].get_value());
```

Comment:
This seems a bit dangerous. The `state_interfaces_` are usually filled in the realtime thread (inside the `update` function). You could put a data structure into a RealtimeBox and fill that in the `update` method (or maybe even the `activate` method would be sufficient?) and read from that box in a safe manner here.

Reply:
I chose to put it in the `on_activate` method. Used a RealtimeBox and a struct to hold the information.